src/shared/local-addresses.c

```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
 
         return (int) n_list;
 }
+
+int local_outbound(sd_netlink *context, int ifindex, int af, struct local_address **ret) {
+        _cleanup_free_ struct local_address *a = NULL, *b = NULL;
+        int r, done_af = AF_UNSPEC, n, found = 0;
+
+        /* Determines our default outbound addresses, i.e. the local address we use to talk to the default
+         * route. This is still an address of the local host (i.e. this doesn't resolve NAT or so), but it's
+         * the one the local IP stack most likely uses to talk to other hosts.
+         *
+         * This works by connect()ing a SOCK_DGRAM socket to the local gateway, and then reading the IP
+         * address off the socket that was chosen for the routing decision.
+         *
+         * This will generate exactly 0, 1 or 2 addresses depending if AF_INET, AF_INET6 is
+         * requested + configured. */
```

Comment:
Hmm, why cut off the other addresses (see "here" below)? In the common scenario, there is just one address of a given family that is preferred, so this doesn't matter. But when we *are* in a setup where there is more than one address, the kernel will "randomly" select one of the routes and use it for a specific destination address. But for another connection to a different target IP it might select a different source address. If we select one of the addresses ourselves, we are not reporting the full information to the user… So this selection needs to be either justified, or dropped.

Reply:
My original usecase for this was to return "*the*" IP address of the local machine, i.e. the primary one that would actually be used for outbound traffic in the likeliest of cases. i.e. it's supposed to answer the question "what is my IP address" as well as we possibly can. 

Now if we'd return a list of multiple addresses here then this would defeat the point partially as the answer we provide would not be "*the*" IP address anymore, but a possible series of them.

Moreover there's a pretty high chance if we'd iterate through all defined gateways we'd always come to the same conclusion, i.e. the same IP address, over and over again, and would then just filter that out again and spend work on it needlessly.

I mean, this is not useful in truly multi-homed systems with multiple default gateways and many local IP addresses anyway, I have no illusions about this. This is supposed to be useful only for those cases where things are simple, i.e. there is one default gateway, and one IP address that is used to talk to it. And these simple cases are probably 99% of the installations I guess, which is good enough for me.

Or to say this differently: if you have multiple default gateways and multiple local IP addresses that are equally suited to talk to them, then `_outbound` is not your thing anyway, and there's value in leaving our search loop early, since this stuff might not be for free if there myriads of souch gateways or IP addresses defined.

Does that make sense?

I'll add a comment about this to the sources.

Reply:
You make it sound like we could do the right thing in more cases.

> Moreover there's a pretty high chance if we'd iterate through all defined gateways we'd always come to the same conclusion, i.e. the same IP address, over and over again, and would then just filter that out again and spend work on it needlessly.

Well, sure, but how much that "extra work" will be? Except very rare cases, people have just a small handful of routes, so the extra connect() is completely negligible.

And then, if we get the same answer in all cases, that's great. We return a single address and the user is happy.
OTOH, if there *are* multiple addresses, and we return multiple addresses, the user is happy again, because they immediately see that there multiple default routes and potentially avoid major confusion. In both cases, cutting the list early sounds like premature optimization. I would be very surprised if the time we save was measurable. I would very much prefer to return a well-defined answer rather than a quick one.

(For me, the special names like `_gateway` are as often useful as a diagnostic tool, if not more often, then a mechanism to make the actual connection. Thus, complete output is more important than output that allows a connection to be made.)